daemon/config/features.go

```diff
@@ -0,0 +1,37 @@
+package config
+
+import "errors"
+
+// DefaultFeatures maps features the engine supports to
+// their default value.
+// When a features becomes on by default, update this.
+var DefaultFeatures = map[string]bool{
+	"containerd-snapshotter": false,
+}
+
+// PopulateFeatures adds all the features the engine is aware of
+// to features, set to their current default value.
+//
+// This is used to populate the daemon config features map, so that
+// instead of this map not including features that are off by default
+// and not enabled – or on by default and not disabled – it contains
+// all features the engine is are of.
+// This prevents older clients from assuming that engine features are
+// disabled when communicating with newer daemons where the feature is
+// now enabled by default.
+func PopulateFeatures(features map[string]bool) error {
+	if features == nil {
+		return errors.New("features cannot be nil")
+	}
+	if DefaultFeatures == nil {
+		return errors.New("DefaultFeatures cannot be nil")
+	}
+
+	for k, v := range DefaultFeatures {
+		if _, exists := features[k]; !exists {
+			features[k] = v
+		}
+	}
+
+	return nil
+}
```

Comment:
I'm still on the fence on whether we should directly expose the features as defined in daemon's config.

In my view, they are more of a "feature flags", rather than concrete features.

I think it would be better to have some granular "capabilities" that _could_ be derived from feature flag.
Like, currently the containerd feature is mostly about providing a multi-platform image store.
However, in future we might have other containerd-specific features - like, let's say support for OCI artifacts.

In this case, the `containerd-snapshotter` feature itself doesn't provide the client with information whether the OCI artifacts are supported or not (without correlating it with a specific API version).

Or, the feature might be implement for containerd integration first, but then also get implemented for graphdrivers. In this case the client would have to be aware of that.

Instead, it would be better if we had a granular "features/capabilities".
For example, considering that containerd integration implements the multi-platform image store AND OCI artifacts, we wouldn't expose the `containerd-snapshotter` as feature, but `multi-platform-images` and `oci-artifacts`.

Reply:
The idea here (and that we discussed at the Thursday call) was to introduce something we can use, now, and that doesn't lock us into anything/prevent us from making changes in the future (`?features=v2`, etc.).

I'm not opposed to a "capabilities" concept + endpoint, API in the future, but I also think trying to implement that all at once right now is complexity for the sake of complexity and it would be better to figure out what we want to improve as we use it and run into it's limitations.

> Or, the feature might be implement for containerd integration first, but then also get implemented for graphdrivers. In this case the client would have to be aware of that.

This is a fair point – but I don't think it's that bad. If that happens, newer clients won't be affected if talking to a graphdrivers daemon, since after negotiating the API version, before making the `/images/create` endpoint call, they can figure out "if API version < x, then `containerd-snapshotter` needs to be enabled. If API version > x, then it's supported for both".

The only unsupported case would be older clients talking to newer daemons that aren't aware that it's now supported for graphdrivers, but older clients not being able to some new feature isn't surprising to me.

Plus, from what we discussed, it's still useful for some providers (Mirantis, DD) to configure their daemons to advertise specific features. As such, exposing the daemon's config `features` is still useful, for other to configure their daemons to advertise whatever custom feature.

The alternative to this is to implement a full on features API, which we can do anyway iterating on this in the future.